interface/wx/translation.h

```diff
@@ -476,6 +476,43 @@ class wxMsgCatalog
  */
 #define wxGETTEXT_IN_CONTEXT_PLURAL(context, string, plural, n)
 
+/**
+    Similar to wxGETTEXT_IN_CONTEXT() but automatically prepends the current
+    platform-specific prefix to the context.
+
+    This macro is useful for translating the same string differently under
+    different platforms. For example, wxWidgets itself uses it to translate the
+    "Help" menu title to just "?" in Italian, but only under MSW, as this
+    convention only exists under this platform, using the following lines in
+    the message catalog file:
+    @code
+    msgid "&Help"
+    msgstr "&Aiuto"
+
+    msgctxt "msw.menu"
+    msgid "&Help"
+    msgstr "&?"
+    @endcode
+
+    meaning that
+    @code
+    menubar->Append(helpMenu, wxGETTEXT_IN_PLATFORM_CONTEXT("menu", "&Help"));
+    @endcode
+
+    will correctly use "?" as translation only under MSW.
+
+    Other supported prefixes are @c "mac." and @c "unix.", see wxPlatformId
+    for more details.
+
+    @see wxGETTEXT_IN_CONTEXT(), wxGetTranslation()
+
+    Note that using this macro requires including both @c wx/intl.h and @c
+    wx/platinfo.h headers.
+
+    @since 3.1.5
+ */
+#define wxGETTEXT_IN_PLATFORM_CONTEXT(context, string)
```

Comment:
There is a big problem with this approach: `xgettext` will only see the literal value of the `context` argument. If `xgettext` was run on the code sample above, it would produce an entry with `msgctxt "menu"`, yet at runtime an entry with context `"msw.menu"` would be searched for, and that wouldn't be a match.

Reply:
I don't see how could this be avoided. The (very) few cases when platform-specific translations are needed would have to be handled manually, i.e. by editing the context in the `.po` file.

What else could we possibly do?

Reply:
Any manual modifications to `.po` files will be commented out by the next run of `msgmerge`. So the modifications would need to be done to the `wxstd.pot` file really. This is doable within wxWidgets's own translations, as part of `wxstd.pot` generation -- but on the other hand, if indeed there are very few cases within wxWidgets itself, wouldn't it be simpler to just explicitly call the `wxGETTEXT_IN_CONTEXT` macro once for each three platforms? Or call a context-aware no-op macro similar to `wxTRANSLATE`.

From the perspective of library user this manual editing of `.po` or `.pot` is too complicated, IMHO. It would require an unusual additional step to the `xgettext` extraction process, and if omitted, there would be no error at any point, but messages with platform-specific context would remain untranslated for no obvious reason. Again, explicitly listing the three possible platforms could be a solution more easy to grasp.

Reply:
> Any manual modifications to `.po` files will be commented out by the next run of `msgmerge`.

Oops, yes, you're right, this is indeed a problem, thanks for pointing it out.

> So the modifications would need to be done to the `wxstd.pot` file really. This is doable within wxWidgets's own translations, as part of `wxstd.pot` generation -- but on the other hand, if indeed there are very few cases within wxWidgets itself, wouldn't it be simpler to just explicitly call the `wxGETTEXT_IN_CONTEXT` macro once for each three platforms? Or call a context-aware no-op macro similar to `wxTRANSLATE`.

Sorry, I'm not sure about what exactly do you propose. Could you please show how would this work?

I.e. we could add `wxTRANSLATE_IN_CONTEXT()`, but how would it be used? We can't write
```cpp
wxTRANSLATE_IN_CONTEXT("mac", "foo")
wxTRANSLATE_IN_CONTEXT("msw", "foo")
wxTRANSLATE_IN_CONTEXT("unix", "foo")
```
for each string...

> From the perspective of library user this manual editing of `.po` or `.pot` is too complicated, IMHO.

Yes, I have to agree with this.

Finally, it seems using contexts for this might not have been a good idea. At least it definitely would be much simpler to just make a special exception for `wxID_HELP` in `wxGetStockLabel()` rather than trying to handle this at translations level.

Or do you (or anybody else) still see some way to make the idea of using contexts work?

Reply:
> > but on the other hand, if indeed there are very few cases within wxWidgets itself, wouldn't it be simpler to just explicitly call the `wxGETTEXT_IN_CONTEXT` macro once for each three platforms? Or call a context-aware no-op macro similar to `wxTRANSLATE`.
> 
> Sorry, I'm not sure about what exactly do you propose. Could you please show how would this work?
> 
> I.e. we could add `wxTRANSLATE_IN_CONTEXT()`, but how would it be used? We can't write
> 
> ```c++
> wxTRANSLATE_IN_CONTEXT("mac", "foo")
> wxTRANSLATE_IN_CONTEXT("msw", "foo")
> wxTRANSLATE_IN_CONTEXT("unix", "foo")
> ```
> 
> for each string...

That's pretty much what I was proposing, and whether such can be written for each string, depends on how many platform-specific strings there are in the first place. So far there has been exactly one, the `"&Help"` one. It wouldn't be unrealistic to write those three lines for `"&Help"` _once_ in the wx source code. Keeping in mind that those do not need to be repeated with every call to `wxGETTEXT_IN_PLATFORM_CONTEXT`, just have them once, somewhere.

For a handful strings maintaining such lists manually would still be feasible...

> Finally, it seems using contexts for this might not have been a good idea. At least it definitely would be much simpler to just make a special exception for `wxID_HELP` in `wxGetStockLabel()` rather than trying to handle this at translations level.

This might be true with strings in the generic or common parts of the source. But in e.g. `src/osx/cocoa/menu.mm` and `src/osx/menu_osx.cpp` there are many platform-specific strings that could be contextized using just the existing macros, as their platform is fixed.

> Or do you (or anybody else) still see some way to make the idea of using contexts work?

I'll keep thinking about options...

Reply:
My idea was to automatically pick up the translations in `$platform.menu` context for the stock labels if they are present. Which already doesn't seem really practical, i.e. we're not going to have `2*3*72` string (see `wxGetStockLabel()`).

It's true that using contexts could still be useful for the strings in platform-specific code, but then we don't need to use `wxPlatformId` and can just hardcode the context, so I still don't see any real use for `wxGETTEXT_IN_PLATFORM_CONTEXT()`...

Reply:
Thanks @lanurmi for saying what I would say, I agree 100% with everything.

`#ifdef` with an explanatory `// TRANSLATORS:` comment seems appropriate for handling of the *Help* string — it's unlikely anything else of this kind will be required.

Reply:
@vslavik Do you also think that `wxGETTEXT_IN_PLATFORM_CONTEXT` can still be useful? AFAICS if there is no way to use it with the standard gettext workflow without hacks, it's just not useful at all, so I'll just drop it, unless I'm missing something.

Reply:
> @vslavik Do you also think that wxGETTEXT_IN_PLATFORM_CONTEXT can still be useful?

I got a race condition with that one. For just one string, used from one place, it's not useful.

It would be potentially useful if we had a few strings like "Help" and/or present in several places in the code. Then you could use `wxGETTEXT_IN_PLATFORM_CONTEXT` at the place of use, avoiding repetitive `#ifdef` logic, and have just one "dummy" place in code where all strings would be manually listed for `xgettext` extraction. It's a similar situation to `wxGetTranslation` (runtime invocation without a string literal) and `wxTRANSLATE` (some placement in code to support extraction).

IOW, I'd say it's something that might be useful in the future (but probably not), but not worth keeping now (not least because it would confuse readers who'd assume it's *needed* for something).